main.go

```diff
@@ -0,0 +1,36 @@
+package main
+
+import (
+	"context"
+
+	"github.com/kcp-dev/apimachinery/pkg/client"
+	"github.com/kcp-dev/logicalcluster/v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+)
+
+func main() {
+	// create an un-scoped client
+	var cfg *rest.Config // loaded somehow
+	unscopedClient, err := client.NewForConfig(cfg)
+	if err != nil {
+		panic(err)
+	}
+
+	// scope the whole group to a cluster, so it can be used in a single-cluster context
+	rbacClient := unscopedClient.Cluster(logicalcluster.New("legacy"))
+	// now, use it as normal
+	rbacClient.RoleBindings("default").Get(context.TODO(), "system", metav1.GetOptions{})
+
+	// defer scoping until later, keeping un-scoped but strongly typed clients for specific resources
+	clusterRoleClient := unscopedClient.ClusterRoles()
+	roleClient := unscopedClient.Roles()
+	// use them *almost* as normal
+	clusterRoleClient.Cluster(logicalcluster.New("whoa")).ClusterRoles().Get(context.TODO(), "foo", metav1.GetOptions{})
+	// NOTE: for cluster-scoped resources we could technically make the above call not stutter on e.g. ClusterRoles():
+	// clusterRoleClient.Cluster(logicalcluster.New("whoa")).Get(context.TODO(), "foo", metav1.GetOptions{})
+
+	roleClient.Cluster(logicalcluster.New("dang")).Roles("default").Get(context.TODO(), "bar", metav1.GetOptions{})
+	// NOTE: for namespace-scoped resources we could technically make the above call not stutter on e.g. Roles():
+	// roleClient.Cluster(logicalcluster.New("dang")).Namespace("default").Get(context.TODO(), "foo", metav1.GetOptions{})
```

Comment:
Although this looks really good this is not how a standard client-go interface looks like, right?

A standard client-go interface exposes `RolesGetter` that allows for `Roles(namespace)` not for `Namespace(namespace)`. It seems we should conform to the standard interface.

Reply:
If we are _not_ trying to be interface-compatible with existing generated clientsets, I think we should prioritize a DX that "makes sense", i.e. `.Namespace(ns)` >>> `.Roles(ns)`

Reply:
Oh that's not what I was suggesting. Huge +1 to cluster, then ns

Reply:
Even if we managed to be interface-compatible for the initial `.ConfigMaps(ns)` call, we'd have to not be as it must be followed by a `.Cluster()` call before using e.g. `.Get()`, so I'm not sure compatibility is critical as it's going to require a refactor anyway.

Reply:
Interface compat is only required if you are unwilling or unable to modify existing code. The latter is e.g. if you're using a 3rd party library that uses generated clientsets... we should probably make sure we don't lose sight of this case.

Reply:
Yep, that case I think is captured well in the design doc. To my eyes we should be able to support it by ensuring that we emit the underlying (`client-gen`) types after cluster scoping, no? Then a multi-cluster context can be bolted on top of an existing controller (and they would run one controller per lcluster?)